src/pipeline-provider.js

```diff
@@ -43,34 +58,55 @@ export class PipelineProvider {
     return pipeline;
   }
 
+  _findStep( name:string ) {
+    return this.steps.find( x => x.slotName === name || x.slotAlias === name );
+  }
+
   /**
   * Adds a step into the pipeline at a known slot location.
   */
   addStep(name: string, step: PipelineStep): void {
-    let found = this.steps.find(x => x.slotName === name || x.slotAlias === name);
+    let found = this._findStep(name);
     if (found) {
-      found.steps.push(step);
+      if ( !found.steps.includes(step) ) { // prevent duplicates
+        found.steps.push(step);
+      }
     } else {
       throw new Error(`Invalid pipeline slot name: ${name}.`);
     }
   }
 
-  _createPipelineSlot(name, alias) {
-    class PipelineSlot {
-      static inject = [Container];
-      static slotName = name;
-      static slotAlias = alias;
-      static steps = [];
-
-      constructor(container) {
-        this.container = container;
-      }
+  /**
+   * Removes a step from a slot in the pipeline
+   */
+  removeStep( name:string, step: PipelineStep ) {
```

Comment:
If you could match the existing style by using a single space after colons, and no space inside parens that would be great.